Unused temporary for lonely vararg in named args rewrite

## reproduction steps

```scala
object t7 {
  def f(a: Int, b: String*) = "first"
}
object Test extends App {
  println(t7.f(1))
  println(t7.f(a = 19))
  println(t7.f(a = 28, b = "sl19"))
  println(t7.f(b = "sl19", a = 28))
}
```

## problem

```scala
$ scalac -Xlint unused-fresh.scala
unused-fresh.scala:8: warning: local val x$1 in value <local Test> is never used
  println(t7.f(b = "sl19", a = 28))
                   ^
one warning found
```
where
```
      scala.Predef.println({
        <artifact> val x$1: Seq = scala.collection.immutable.Seq.apply(scala.runtime.ScalaRunTime.wrapRefArray(Array[String]{"sl19"}.$asInstanceOf[Array[Object]]())).$asInstanceOf[Seq]();
        <artifact> val x$2: Int(28) = 28;
        t7.f(28, scala.runtime.ScalaRunTime.wrapRefArray(Array[String]{"sl19"}.$asInstanceOf[Array[Object]]()))
      });
```

## expectation

Less cruft.

Another version, where even a simple usage of the varargs emits the warning:

## reproduction steps

using Scala 2.13.3

```scala
object Hmm {
  def zxc(b: Int*)(implicit x: Int = 3) = "" + b + x
  zxc(4)
}
```

## problem

```
> scalac -Xlint as.scala
as.scala:3: warning: local val x$1 in value <local Hmm> is never used
  zxc(4)
      ^
1 warning
```

I ran into this one as well. Minimized to:

```scala
cat sandbox/test.scala && scalac --scala-version 2.13.6 -Werror -Xlint:unused -d /tmp -Xprint:typer sandbox/test.scala
class Test {

   def foo(a: Any, b: Any = null, c: Any = null)(cs: String*) = ???
   foo("", c = "")("X")
}
sandbox/test.scala:4: warning: local val x$4 in value <local Test> is never used
   foo("", c = "")("X")
                   ^
[[syntax trees at end of                     typer]] // test.scala
package <empty> {
  class Test extends scala.AnyRef {
    def <init>(): Test = {
      Test.super.<init>();
      ()
    };
    def foo(a: Any, b: Any = null, c: Any = null)(cs: String*): Nothing = scala.Predef.???;
    <synthetic> def foo$default$2: Any = null;
    <synthetic> def foo$default$3: Any = null;
    {
      <artifact> val x$1: String("") = "";
      <artifact> val x$2: String("") = "";
      <artifact> val x$3: Any = Test.this.foo$default$2;
      <artifact> val x$4: Seq[String("X")] @scala.reflect.internal.annotations.uncheckedBounds = scala.collection.immutable.Seq.apply[String]("X");
      Test.this.foo("", x$3, "")("X")
    }
  }
}

error: No warnings can be incurred under -Werror.
1 warning
1 error
```

--- a/src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala
@@ -302,18 +302,15 @@ trait NamesDefaults { self: Analyzer =>
         case _ =>
           val byName   = isByNameParamType(paramTpe)
           val repeated = isScalaRepeatedParamType(paramTpe)
-          val argTpe = (
-            if (repeated) arg match {
+          // TODO In 83c9c764b, we tried to a stable type here to fix scala/bug#7234. But the resulting TypeTree over a
+          //      singleton type without an original TypeTree fails to retypecheck after a resetAttrs (scala/bug#7516),
+          //      which is important for (at least) macros.
+          val argTpe =
+            arg match {
+              case _ if !repeated        => arg.tpe
               case WildcardStarArg(expr) => expr.tpe
-              case _                     => seqType(arg.tpe)
+              case _                     => seqType(arg.tpe.widen)      // avoid constant type
             }
-            else {
-              // TODO In 83c9c764b, we tried to a stable type here to fix scala/bug#7234. But the resulting TypeTree over a
-              //      singleton type without an original TypeTree fails to retypecheck after a resetAttrs (scala/bug#7516),
-              //      which is important for (at least) macros.
-              arg.tpe
-            }
-          )
           val s = context.owner.newValue(freshTermName(nme.NAMEDARG_PREFIX)(typer.fresh), arg.pos, newFlags = ARTIFACT) setInfo {
             val tp = if (byName) functionType(Nil, argTpe) else argTpe
             uncheckedBounds(tp)
@@ -330,10 +327,11 @@ trait NamesDefaults { self: Analyzer =>
               res
             } else {
               new ChangeOwnerTraverser(context.owner, sym) traverse arg // fixes #4502
-              if (repeated) arg match {
+              arg match {
+                case _ if !repeated        => arg
                 case WildcardStarArg(expr) => expr
-                case _                     => blockTyper typed gen.mkSeqApply(resetAttrs(arg))
-              } else arg
+                case _                     => blockTyper.typed(gen.mkSeqApply(resetAttrs(arg)))
+              }
             }
           Some(atPos(body.pos)(ValDef(sym, body).setType(NoType)))
       }

--- a/test/files/pos/t11964.scala
+++ b/test/files/pos/t11964.scala
@@ -1,0 +1,19 @@
+// scalac: -Werror -Xlint
+
+object Hmm {
+  def zxc(b: Int*)(implicit x: Int = 3) = "" + b + x
+  def res = zxc(4)
+}
+
+object Test {
+  def foo(a: Any, b: Any = null, c: Any = null)(cs: String*) = ???
+  def res = foo("", c = "")("X")
+}
+
+object OP {
+  def f(a: Int, b: String*) = "first"
+  def res = f(b = "sl19", a = 28)  // looks like the issue is only with single arg supplied to varargs.
+  def or  = f(b = ("x"::"y"::Nil):_*, a = 42)  // 2.13 syntax only
+  //def and = f(b = ("x"::"y"::Nil):_*)   // broken under 2.13, which disallows default + varargs
+  def and = List(elems = ("x"::"y"::Nil):_*)
+}

--- a/test/files/run/names-defaults.check
+++ b/test/files/run/names-defaults.check
@@ -13,9 +13,6 @@ names-defaults.scala:371: warning: the parameter name x is deprecated: use s ins
 names-defaults.scala:35: warning: local var var2 in value <local Test> is never used
     var var2 = 0
         ^
-names-defaults.scala:108: warning: local val x$34 in value <local Test> is never used
-  println(t7.f(b = "sl19", a = 28)) // first
-                   ^
 names-defaults.scala:279: warning: local val u in method foo is never used
   class A2489 { def foo(): Unit = { def bar(a: Int = 1) = a; bar(); val u = 0 } }
                                                                         ^
@@ -25,12 +22,6 @@ names-defaults.scala:280: warning: local val v in method foo is never used
 names-defaults.scala:280: warning: local val u in method foo is never used
   class A2489x2 { def foo(): Unit = { val v = 10; def bar(a: Int = 1, b: Int = 2) = a; bar(); val u = 0 } }
                                                                                                   ^
-names-defaults.scala:380: warning: local val x$104 in value <local Test> is never used
-  println(t3697.a(3)())
-                  ^
-names-defaults.scala:385: warning: local val x$112 in value <local Test> is never used
-  println(t3697.b(b = 1, a = 2, c = 3))
-                                    ^
 names-defaults.scala:269: warning: multiline expressions might require enclosing parentheses; a value can be silently discarded when Unit is expected
     spawn(b = { val ttt = 1; ttt }, a = 0)
                              ^

